source/common/router/config_impl.cc

```diff
@@ -1103,33 +1103,63 @@ RouteMatcher::RouteMatcher(const envoy::config::route::v3::RouteConfiguration& r
   }
 }
 
-RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const Http::RequestHeaderMap& headers,
+RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const RouteCallback& cb,
+                                                         const Http::RequestHeaderMap& headers,
                                                          const StreamInfo::StreamInfo& stream_info,
                                                          uint64_t random_value) const {
   // No x-forwarded-proto header. This normally only happens when ActiveStream::decodeHeaders
   // bails early (as it rejects a request), so there is no routing is going to happen anyway.
   const auto* forwarded_proto_header = headers.ForwardedProto();
   if (forwarded_proto_header == nullptr) {
+    if (cb) {
+      cb(nullptr, RouteEvalStatus::NoMoreRoutes);
+    }
     return nullptr;
   }
 
   // First check for ssl redirect.
   if (ssl_requirements_ == SslRequirements::All && forwarded_proto_header->value() != "https") {
+    if (cb) {
+      cb(SSL_REDIRECT_ROUTE, RouteEvalStatus::NoMoreRoutes);
+    }
     return SSL_REDIRECT_ROUTE;
   } else if (ssl_requirements_ == SslRequirements::ExternalOnly &&
              forwarded_proto_header->value() != "https" &&
              !Http::HeaderUtility::isEnvoyInternalRequest(headers)) {
+    if (cb) {
+      cb(SSL_REDIRECT_ROUTE, RouteEvalStatus::NoMoreRoutes);
+    }
     return SSL_REDIRECT_ROUTE;
   }
 
   // Check for a route that matches the request.
-  for (const RouteEntryImplBaseConstSharedPtr& route : routes_) {
-    RouteConstSharedPtr route_entry = route->matches(headers, stream_info, random_value);
-    if (nullptr != route_entry) {
-      return route_entry;
+  for (auto route = routes_.begin(); route != routes_.end(); ++route) {
+    RouteConstSharedPtr route_entry = (*route)->matches(headers, stream_info, random_value);
+    if (nullptr == route_entry) {
+      continue;
     }
+
+    if (cb) {
+      RouteEvalStatus eval_status = (std::next(route) == routes_.end())
+                                        ? RouteEvalStatus::NoMoreRoutes
+                                        : RouteEvalStatus::HasMoreRoutes;
+      RouteMatchStatus match_status = cb(route_entry, eval_status);
+      if (match_status == RouteMatchStatus::Accept) {
+        return route_entry;
+      }
+      if (match_status == RouteMatchStatus::Continue &&
+          eval_status == RouteEvalStatus::NoMoreRoutes) {
+        return nullptr;
+      }
+      continue;
+    }
+
+    return route_entry;
   }
 
+  if (cb) {
+    cb(nullptr, RouteEvalStatus::NoMoreRoutes);
+  }
```

Comment:
Similar to the case above, do you view this as useful from a callback perspective? I'm wondering if we should just delete this case for now? Same elsewhere. If we want to keep it, I would probably consider making the callbacks an interface, and potentially having a `noRouteError()` or similar callback.

Reply:
Matt, If I think it is better to invoke callback only if a matching route is found and give a chance to override it would simplify the  code as well. I'll amend the documentation as well.